File: GameLibrary/src/cc/lib/zombicide/ZActionType.java
package cc.lib.zombicide;

import cc.lib.annotation.Keep;
import cc.lib.game.Utils;

@Keep
public enum ZActionType {
    NOTHING,
    MOVE,
    SEARCH,
    OPEN_DOOR,
    CLOSE_DOOR,
    MELEE,
    RANGED,
    MAGIC,
    THROW_ITEM,
    ENCHANTMENT,
    ACTIVATE,
    INVENTORY,
    CONSUME,
    OBJECTIVE,
    RELOAD,
    DROP_ITEM,
    PICKUP_ITEM,
    MAKE_NOISE,
    SHOVE,
    DEFEND,
    BEQUEATH_MOVE;

    public boolean oncePerTurn() {
        switch (this) {
            case SEARCH:
            case SHOVE:
            case ENCHANTMENT:
                return true;
        }
        return false;
    }

    public int costPerTurn() {
        return 1;
    }

    public boolean breaksInvisibility() {
        switch (this) {
            case MELEE:
            case RANGED:
            case MAGIC:
            case SHOVE:
            case OPEN_DOOR:
            case CLOSE_DOOR:
            case MAKE_NOISE:
                return true;
        }
        return false;
    }

    public boolean isRanged() {
        switch (this) {
            case RANGED:
                return true;
        }
        return false;
    }

    public boolean isMagic() {
        switch (this) {
            case MAGIC:
                return true;
        }
        return false;
    }

    public boolean isProjectile() {
        switch (this) {
            case RANGED:
            case MAGIC:
            case THROW_ITEM:
                return true;
        }
        return false;
    }

    public boolean isMelee() {
        return this == MELEE;
    }

    public boolean isMovement() {
        return this == MOVE;
    }

    public String getLabel() {
        return Utils.toPrettyString(name());
    }
}

File: GameLibrary/src/cc/lib/zombicide/ZActionType.kt
package cc.lib.zombicide

import cc.lib.annotation.Keep
import cc.lib.utils.prettify

@Keep
enum class ZActionType {
    NOTHING,
    MOVE,
    SEARCH,
    OPEN_DOOR,
    CLOSE_DOOR,
    BARRICADE_DOOR {
        override fun costPerTurn(): Int {
            return 3
        }
    },
    MELEE,
    RANGED,
    MAGIC,
    THROW_ITEM,
    ENCHANTMENT,
    ACTIVATE,
    INVENTORY,
    CONSUME,
    OBJECTIVE,
    RELOAD,
    DROP_ITEM,
    PICKUP_ITEM,
    MAKE_NOISE,
    SHOVE,
    DEFEND,
    BEQUEATH_MOVE;

    fun oncePerTurn(): Boolean {
        when (this) {
            SEARCH, SHOVE, ENCHANTMENT -> return true
        }
        return false
    }

    open fun costPerTurn(): Int {
        return 1
    }

    fun breaksInvisibility(): Boolean {
        when (this) {
            MELEE, RANGED, MAGIC, SHOVE, OPEN_DOOR, CLOSE_DOOR, BARRICADE_DOOR, MAKE_NOISE -> return true
        }
        return false
    }

    val isRanged: Boolean
        get() = this === RANGED
    val isMagic: Boolean
        get() = this === MAGIC
    val isProjectile: Boolean
        get() {
            when (this) {
                RANGED, MAGIC, THROW_ITEM -> return true
            }
            return false
        }
    val isMelee: Boolean
        get() = this === MELEE
    val isMovement: Boolean
        get() = this === MOVE
    val label: String
        get() = prettify(name)
}